[BUG] flaky test cases in component_controller_test.go

**Describe the bug**
https://github.com/apecloud/kubeblocks/actions/runs/11267044653/job/31331526262?pr=8251

```
Running Suite: Controller Suite - /home/runner/work/kubeblocks/kubeblocks/controllers/apps
==========================================================================================
Random Seed: 1728532048

Will run 136 of 136 specs
•••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••
------------------------------
• [FAILED] [11.456 seconds]
Component Controller provisioning [It] with component replicas limit
/home/runner/work/kubeblocks/kubeblocks/controllers/apps/component_controller_test.go:1716

  Timeline >>
  STEP: clean resources @ 10/10/24 03:48:56.602
  STEP: clear resources Cluster @ 10/10/24 03:48:56.602
  STEP: clear resources ClusterDefinition @ 10/10/24 03:48:56.604
  STEP: clear resources ComponentDefinition @ 10/10/24 03:48:56.607
  STEP: clear resources ComponentVersion @ 10/10/24 03:48:56.619
  STEP: clear resources Component @ 10/10/24 03:48:56.622
  STEP: clear resources PersistentVolumeClaim @ 10/10/24 03:48:56.69
  STEP: clear resources Pod @ 10/10/24 03:48:56.695
  STEP: clear resources Backup @ 10/10/24 03:48:56.697
  STEP: clear resources BackupPolicy @ 10/10/24 03:48:56.7
  STEP: clear resources VolumeSnapshot @ 10/10/24 03:48:56.748
  STEP: clear resources BackupPolicyTemplate @ 10/10/24 03:48:56.758
  STEP: clear resources ActionSet @ 10/10/24 03:48:56.76
  STEP: clear resources StorageClass @ 10/10/24 03:48:56.763
  STEP: Create a componentDefinition obj @ 10/10/24 03:48:56.766
  STEP: Create a componentVersion obj @ 10/10/24 03:48:56.779
  STEP: Mock kb-agent client for the default transformer of system accounts provision @ 10/10/24 03:48:56.788
  STEP: create actionSet @ 10/10/24 03:48:56.788
  STEP: Creating a BackupPolicyTemplate @ 10/10/24 03:48:56.798
  STEP: create component w/o replicas limit set @ 10/10/24 03:48:56.81
  STEP: Creating a cluster with new component definition @ 10/10/24 03:48:56.81
  STEP: Waiting for the cluster enter expected phase @ 10/10/24 03:48:56.815
  STEP: Waiting for the component enter expected phase @ 10/10/24 03:48:57.671
  STEP: set replicas limit @ 10/10/24 03:48:57.697
  STEP: create component w/ replicas limit set - out-of-limit @ 10/10/24 03:48:57.716
  STEP: Creating a cluster with new component definition @ 10/10/24 03:48:57.716
  STEP: Waiting for the cluster enter expected phase @ 10/10/24 03:48:57.732
  [FAILED] in [It] - /home/runner/work/kubeblocks/kubeblocks/controllers/apps/component_controller_test.go:189 @ 10/10/24 03:49:07.733
  STEP: clean resources @ 10/10/24 03:49:07.733
  STEP: clear resources Cluster @ 10/10/24 03:49:07.733
  STEP: clear resources ClusterDefinition @ 10/10/24 03:49:07.772
  STEP: clear resources ComponentDefinition @ 10/10/24 03:49:07.776
  STEP: clear resources ComponentVersion @ 10/10/24 03:49:07.813
  STEP: clear resources Component @ 10/10/24 03:49:07.834
  STEP: clear resources PersistentVolumeClaim @ 10/10/24 03:49:07.867
  STEP: clear resources Pod @ 10/10/24 03:49:07.87
  STEP: clear resources Backup @ 10/10/24 03:49:07.873
  STEP: clear resources BackupPolicy @ 10/10/24 03:49:07.875
  STEP: clear resources VolumeSnapshot @ 10/10/24 03:49:07.903
  STEP: clear resources BackupPolicyTemplate @ 10/10/24 03:49:07.907
  STEP: clear resources ActionSet @ 10/10/24 03:49:07.922
  STEP: clear resources StorageClass @ 10/10/24 03:49:07.929
  STEP: clean resources @ 10/10/24 03:49:07.931
  STEP: clear resources Cluster @ 10/10/24 03:49:07.931
  STEP: clear resources ClusterDefinition @ 10/10/24 03:49:07.933
  STEP: clear resources ComponentDefinition @ 10/10/24 03:49:07.935
  STEP: clear resources ComponentVersion @ 10/10/24 03:49:07.937
  STEP: clear resources Component @ 10/10/24 03:49:07.941
  STEP: clear resources PersistentVolumeClaim @ 10/10/24 03:49:08.01
  STEP: clear resources Pod @ 10/10/24 03:49:08.014
  STEP: clear resources Backup @ 10/10/24 03:49:08.018
  STEP: clear resources BackupPolicy @ 10/10/24 03:49:08.022
  STEP: clear resources VolumeSnapshot @ 10/10/24 03:49:08.048
  STEP: clear resources BackupPolicyTemplate @ 10/10/24 03:49:08.05
  STEP: clear resources ActionSet @ 10/10/24 03:49:08.053
  STEP: clear resources StorageClass @ 10/10/24 03:49:08.055
  << Timeline

  [FAILED] Timed out after 10.000s.
  Expected
      <int64>: 2
  to be equivalent to
      <int>: 1
  In [It] at: /home/runner/work/kubeblocks/kubeblocks/controllers/apps/component_controller_test.go:189 @ 10/10/24 03:49:07.733
------------------------------
••••••••••••••••••••••••••••••••••••W1010 03:50:16.067200   30586 reflector.go:462] pkg/mod/k8s.io/client-go@v0.29.2/tools/cache/reflector.go:229: watch of *v1.PersistentVolume ended with: an error on the server ("unable to decode an event from the watch stream: context canceled") has prevented the request from succeeding
W1010 03:50:16.067277   30586 reflector.go:462] pkg/mod/k8s.io/client-go@v0.29.2/tools/cache/reflector.go:229: watch of *v1.ComponentDefinition ended with: an error on the server ("unable to decode an event from the watch stream: context canceled") has prevented the request from succeeding
W1010 03:50:16.067306   30586 reflector.go:462] pkg/mod/k8s.io/client-go@v0.29.2/tools/cache/reflector.go:229: watch of *v1.ConfigMap ended with: an error on the server ("unable to decode an event from the watch stream: context canceled") has prevented the request from succeeding
W1010 03:50:16.067[406](https://github.com/apecloud/kubeblocks/actions/runs/11267044653/job/31331526262?pr=8251#step:6:407)   30586 reflector.go:462] pkg/mod/k8s.io/client-go@v0.29.2/tools/cache/reflector.go:229: watch of *v1.ServiceAccount ended with: an error on the server ("unable to decode an event from the watch stream: context canceled") has prevented the request from succeeding


Summarizing 1 Failure:
  [FAIL] Component Controller provisioning [It] with component replicas limit
  /home/runner/work/kubeblocks/kubeblocks/controllers/apps/component_controller_test.go:189

Ran 136 of 136 Specs in 168.031 seconds
FAIL! -- 135 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestAPIs (168.04s)
&StatefulSet{ObjectMeta:{hello      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []},Spec:StatefulSetSpec{Replicas:nil,Selector:nil,Template:{{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []} {[] [] [] []  <nil> <nil>  map[]   <nil>  false false false <nil> nil []   nil  [] []  <nil> nil [] <nil> <nil> <nil> map[] [] <nil> nil <nil> [] []}},VolumeClaimTemplates:[]PersistentVolumeClaim{},ServiceName:,PodManagementPolicy:,UpdateStrategy:StatefulSetUpdateStrategy{Type:,RollingUpdate:nil,},RevisionHistoryLimit:nil,MinReadySeconds:0,PersistentVolumeClaimRetentionPolicy:nil,Ordinals:nil,},Status:StatefulSetStatus{ObservedGeneration:0,Replicas:0,ReadyReplicas:0,CurrentReplicas:0,UpdatedReplicas:0,CurrentRevision:,UpdateRevision:,CollisionCount:nil,Conditions:[]StatefulSetCondition{},AvailableReplicas:0,},}
{<nil> nil {{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []} {[] [] [] []  <nil> <nil>  map[]   <nil>  false false false <nil> nil []   nil  [] []  <nil> nil [] <nil> <nil> <nil> map[] [] <nil> nil <nil> [] []}} []   { nil} <nil> 0 nil nil}
FAIL
coverage: 75.8% of statements
FAIL	github.com/apecloud/kubeblocks/controllers/apps	168.095s
```

**To Reproduce**
Steps to reproduce the behavior:
1. Go to '...'
2. Click on '....'
3. Scroll down to '....'
4. See error

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Desktop (please complete the following information):**
 - OS: [e.g. iOS]
 - Browser [e.g. chrome, safari]
 - Version [e.g. 22]

**Additional context**
Add any other context about the problem here.

--- a/controllers/apps/cluster_controller_test.go
+++ b/controllers/apps/cluster_controller_test.go
@@ -196,7 +196,7 @@ var _ = Describe("Cluster Controller", func() {
 	}
 
 	waitForCreatingResourceCompletely := func(clusterKey client.ObjectKey, compNames ...string) {
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+		Eventually(testapps.ClusterReconciled(&testCtx, clusterKey)).Should(BeTrue())
 		cluster := &appsv1.Cluster{}
 		Eventually(testapps.CheckObjExists(&testCtx, clusterKey, cluster, true)).Should(Succeed())
 		for _, compName := range compNames {
@@ -260,7 +260,7 @@ var _ = Describe("Cluster Controller", func() {
 		createClusterObjNoWait("", componentProcessorWrapper(compName, compDefName, processor))
 
 		By("Waiting for the cluster enter Creating phase")
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+		Eventually(testapps.ClusterReconciled(&testCtx, clusterKey)).Should(BeTrue())
 		Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(appsv1.CreatingClusterPhase))
 
 		By("Wait component created")
@@ -277,7 +277,7 @@ var _ = Describe("Cluster Controller", func() {
 		createClusterObjNoWait(clusterDefObj.Name, componentProcessorWrapper(compName, "", setTopology, processor))
 
 		By("Waiting for the cluster enter Creating phase")
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+		Eventually(testapps.ClusterReconciled(&testCtx, clusterKey)).Should(BeTrue())
 		Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(appsv1.CreatingClusterPhase))
 
 		By("Wait components created")
@@ -293,7 +293,7 @@ var _ = Describe("Cluster Controller", func() {
 		createClusterObjNoWait("", shardingComponentProcessorWrapper(compTplName, compDefName, processor))
 
 		By("Waiting for the cluster enter Creating phase")
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+		Eventually(testapps.ClusterReconciled(&testCtx, clusterKey)).Should(BeTrue())
 		Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(appsv1.CreatingClusterPhase))
 
 		By("Wait component created")
@@ -323,7 +323,7 @@ var _ = Describe("Cluster Controller", func() {
 		createClusterObjNoWait("", multipleTemplateComponentProcessorWrapper(compName, compDefName, processor))
 
 		By("Waiting for the cluster enter Creating phase")
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+		Eventually(testapps.ClusterReconciled(&testCtx, clusterKey)).Should(BeTrue())
 		Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(appsv1.CreatingClusterPhase))
 
 		By("Wait component created")

--- a/controllers/apps/component_controller_test.go
+++ b/controllers/apps/component_controller_test.go
@@ -159,7 +159,7 @@ var _ = Describe("Component Controller", func() {
 	}
 
 	waitForCreatingResourceCompletely := func(clusterKey client.ObjectKey, compNames ...string) {
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+		Eventually(testapps.ClusterReconciled(&testCtx, clusterKey)).Should(BeTrue())
 		cluster := &kbappsv1.Cluster{}
 		Eventually(testapps.CheckObjExists(&testCtx, clusterKey, cluster, true)).Should(Succeed())
 		for _, compName := range compNames {
@@ -186,7 +186,7 @@ var _ = Describe("Component Controller", func() {
 		clusterKey = client.ObjectKeyFromObject(clusterObj)
 
 		By("Waiting for the cluster enter expected phase")
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+		Eventually(testapps.ClusterReconciled(&testCtx, clusterKey)).Should(BeTrue())
 		if phase == nil {
 			Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(kbappsv1.CreatingClusterPhase))
 		} else {
@@ -201,7 +201,7 @@ var _ = Describe("Component Controller", func() {
 		compObj = &kbappsv1.Component{}
 		Eventually(testapps.CheckObjExists(&testCtx, compKey, compObj, true)).Should(Succeed())
 		if phase == nil {
-			Eventually(testapps.GetComponentObservedGeneration(&testCtx, compKey)).Should(BeEquivalentTo(1))
+			Eventually(testapps.ComponentReconciled(&testCtx, compKey)).Should(BeTrue())
 			Eventually(testapps.GetComponentPhase(&testCtx, compKey)).Should(Equal(kbappsv1.CreatingClusterCompPhase))
 		}
 	}
@@ -264,6 +264,28 @@ var _ = Describe("Component Controller", func() {
 			Expect(cluster.Spec.ComponentSpecs).Should(HaveLen(1))
 			cluster.Spec.ComponentSpecs[0].Replicas = replicas
 		})()).ShouldNot(HaveOccurred())
+	}
+
+	getStableClusterObservedGeneration := func(clusterKey types.NamespacedName, waitFor *time.Duration) (int64, *kbappsv1.Cluster) {
+		sleepTime := 300 * time.Millisecond
+		if waitFor != nil {
+			sleepTime = *waitFor
+		}
+		time.Sleep(sleepTime)
+		cluster := &kbappsv1.Cluster{}
+		Expect(testCtx.Cli.Get(testCtx.Ctx, clusterKey, cluster)).Should(Succeed())
+		return cluster.Status.ObservedGeneration, cluster
+	}
+
+	getStableComponentObservedGeneration := func(compKey types.NamespacedName, waitFor *time.Duration) (int64, *kbappsv1.Component) {
+		sleepTime := 300 * time.Millisecond
+		if waitFor != nil {
+			sleepTime = *waitFor
+		}
+		time.Sleep(sleepTime)
+		comp := &kbappsv1.Component{}
+		Expect(testCtx.Cli.Get(testCtx.Ctx, compKey, comp)).Should(Succeed())
+		return comp.Status.ObservedGeneration, comp
 	}
 
 	testChangeReplicas := func(compName, compDefName string) {
@@ -633,9 +655,7 @@ var _ = Describe("Component Controller", func() {
 	horizontalScale := func(updatedReplicas int, storageClassName string, policyType *string, compDefNames ...string) {
 		defer kbagent.UnsetMockClient()
 
-		cluster := &kbappsv1.Cluster{}
-		Expect(k8sClient.Get(testCtx.Ctx, clusterKey, cluster)).Should(Succeed())
-		initialGeneration := int(cluster.Status.ObservedGeneration)
+		initialGeneration, cluster := getStableClusterObservedGeneration(clusterKey, nil)
 
 		setHorizontalScalePolicy(policyType, compDefNames...)
 
@@ -665,7 +685,7 @@ var _ = Describe("Component Controller", func() {
 
 		By("Checking cluster status and the number of replicas changed")
 		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).
-			Should(BeEquivalentTo(initialGeneration + len(cluster.Spec.ComponentSpecs)))
+			Should(BeEquivalentTo(int(initialGeneration) + len(cluster.Spec.ComponentSpecs)))
 	}
 
 	testHorizontalScale := func(compName, compDefName string, initialReplicas, updatedReplicas int32, bpt *string) {
@@ -743,7 +763,7 @@ var _ = Describe("Component Controller", func() {
 			testk8s.MockInstanceSetReady(its, mockPods...)
 		})).ShouldNot(HaveOccurred())
 
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+		initialGeneration, _ := getStableClusterObservedGeneration(clusterKey, nil)
 		Eventually(testapps.GetClusterComponentPhase(&testCtx, clusterKey, compName)).Should(Equal(kbappsv1.RunningClusterCompPhase))
 		Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(kbappsv1.RunningClusterPhase))
 
@@ -758,7 +778,7 @@ var _ = Describe("Component Controller", func() {
 		})()).ShouldNot(HaveOccurred())
 
 		By("Checking the resize operation in progress for data volume")
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(2))
+		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(initialGeneration + 1))
 		Eventually(testapps.GetClusterComponentPhase(&testCtx, clusterKey, compName)).Should(Equal(kbappsv1.UpdatingClusterCompPhase))
 		Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(kbappsv1.UpdatingClusterPhase))
 		for i := 0; i < replicas; i++ {
@@ -789,7 +809,7 @@ var _ = Describe("Component Controller", func() {
 		Expect(testapps.GetAndChangeObjStatus(&testCtx, client.ObjectKeyFromObject(its), func(its *workloads.InstanceSet) {
 			testk8s.MockInstanceSetReady(its, mockPods...)
 		})()).ShouldNot(HaveOccurred())
-		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(2))
+		Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(initialGeneration + 1))
 		Eventually(testapps.GetClusterComponentPhase(&testCtx, clusterKey, compName)).Should(Equal(kbappsv1.RunningClusterCompPhase))
 		Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(kbappsv1.RunningClusterPhase))
 
@@ -903,9 +923,12 @@ var _ = Describe("Component Controller", func() {
 			}
 		}
 
-		checkResizeOperationFinished := func(generation int64) {
-			Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(generation))
-			Eventually(testapps.GetComponentObservedGeneration(&testCtx, compKey)).Should(BeEquivalentTo(generation))
+		initialClusterGeneration, _ := getStableClusterObservedGeneration(clusterKey, nil)
+		initialComponentGeneration, _ := getStableComponentObservedGeneration(compKey, pointer.Duration(0) /* no need to sleep */)
+
+		checkResizeOperationFinished := func(diffGeneration int64) {
+			Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(initialClusterGeneration + diffGeneration))
+			Eventually(testapps.GetComponentObservedGeneration(&testCtx, compKey)).Should(BeEquivalentTo(initialComponentGeneration + diffGeneration))
 		}
 
 		By("Updating the PVC storage size")
@@ -913,7 +936,7 @@ var _ = Describe("Component Controller", func() {
 		changePVC(newStorageValue)
 
 		By("Checking the resize operation finished")
-		checkResizeOperationFinished(2)
+		checkResizeOperationFinished(1)
 
 		By("Checking PVCs are resized")
 		checkPVC(newStorageValue)
@@ -923,7 +946,7 @@ var _ = Describe("Component Controller", func() {
 		changePVC(originStorageValue)
 
 		By("Checking the resize operation finished")
-		checkResizeOperationFinished(3)
+		checkResizeOperationFinished(2)
 
 		By("Checking PVCs are resized")
 		checkPVC(originStorageValue)

--- a/controllers/apps/transformer_component_tls_test.go
+++ b/controllers/apps/transformer_component_tls_test.go
@@ -170,7 +170,7 @@ var _ = Describe("TLS self-signed cert function", func() {
 					GetObject()
 				clusterKey := client.ObjectKeyFromObject(clusterObj)
 				Eventually(k8sClient.Get(ctx, clusterKey, clusterObj)).Should(Succeed())
-				Eventually(testapps.GetClusterObservedGeneration(&testCtx, clusterKey)).Should(BeEquivalentTo(1))
+				Eventually(testapps.ClusterReconciled(&testCtx, clusterKey)).Should(BeTrue())
 				Eventually(testapps.GetClusterPhase(&testCtx, clusterKey)).Should(Equal(appsv1.CreatingClusterPhase))
 
 				itsList := testk8s.ListAndCheckInstanceSet(&testCtx, clusterKey)

--- a/pkg/testutil/apps/cluster_util.go
+++ b/pkg/testutil/apps/cluster_util.go
@@ -85,6 +85,15 @@ func GetClusterObservedGeneration(testCtx *testutil.TestContext, clusterKey type
 	}
 }
 
+// ClusterReconciled checks if the testing cluster has been reconciled.
+func ClusterReconciled(testCtx *testutil.TestContext, clusterKey types.NamespacedName) func(gomega.Gomega) bool {
+	return func(g gomega.Gomega) bool {
+		cluster := &appsv1.Cluster{}
+		g.Expect(testCtx.Cli.Get(testCtx.Ctx, clusterKey, cluster)).Should(gomega.Succeed())
+		return cluster.Status.ObservedGeneration > 0 && cluster.Status.ObservedGeneration == cluster.Generation
+	}
+}
+
 // NewPVCSpec creates appsv1alpha1.PersistentVolumeClaimSpec.
 func NewPVCSpec(size string) appsv1.PersistentVolumeClaimSpec {
 	return appsv1.PersistentVolumeClaimSpec{

--- a/pkg/testutil/apps/component_util.go
+++ b/pkg/testutil/apps/component_util.go
@@ -44,3 +44,12 @@ func GetComponentPhase(testCtx *testutil.TestContext, compKey types.NamespacedNa
 		return comp.Status.Phase
 	}
 }
+
+// ComponentReconciled checks if the testing component has been reconciled.
+func ComponentReconciled(testCtx *testutil.TestContext, compKey types.NamespacedName) func(gomega.Gomega) bool {
+	return func(g gomega.Gomega) bool {
+		comp := &appsv1.Component{}
+		g.Expect(testCtx.Cli.Get(testCtx.Ctx, compKey, comp)).Should(gomega.Succeed())
+		return comp.Status.ObservedGeneration > 0 && comp.Status.ObservedGeneration == comp.Generation
+	}
+}

